### PR TITLE
Mod activity

### PR DESCRIPTION
**Pull request by HitoriWoB**
Add 3-day to all ban types
Add 30-minute and 60-minute ban types to Chat and PM bans
Added basic moderator stats tool to be able to review staff actions
Added detailed data for in-depth reviews

### CODE REVIEW

File: classes/StaffManager.php [hunk 1]
```diff
@@ -513,60 +517,77 @@ public function canIssueOW($staff_level) {
...
+        if($ban_length == self::PERM_BAN_VALUE) {
+            $ban_expire = $ban_length;
+        }
+        else {
+            // Bans are reported in minutes, add second value to current time
+            $ban_expire = time() + ($ban_length * 60);
+        }
+
+        //Create new ban data for storage
+        if($ban_data == null || empty($ban_data)) {
```
Comment (levimeahan) [inline classes/StaffManager.php:532]: I think these two are redundant, can probably just do
```suggestion
        if(empty($ban_data)) {
```

File: classes/StaffManager.php [hunk 2]
```diff
@@ -513,60 +517,77 @@ public function canIssueOW($staff_level) {
...
-            $ban_data = json_decode($user_data['ban_data'], true);
-            $ban_expire = ($ban_length == self::PERM_BAN_VALUE ? $ban_length : time() + ($ban_length * 86400));
-            $old_ban_remaining = false;
-
-            //Create new ban data for storage
-            if($ban_data == null || empty($ban_data)) {
-                $ban_data = json_encode(array(
-                    $ban_type => $ban_expire
-                ));
+        $ban_data = json_decode($user_data['ban_data'], true);
```
Comment (levimeahan) [inline classes/StaffManager.php:520]: Follow-up request, can we get a DTO for ban data with a static method to initialize from db data?

File: pages/modPanel.php [hunk 3]
```diff
@@ -671,6 +677,162 @@ function modPanel() {
...
+                }
+                if(in_array($_POST['start_month'], ['Apr', 'Jun', 'Sep', 'Nov'])) {
+                    $start_day = min($start_day, 30);
+                }
+                if(in_array($_POST['end_month'], ['Apr', 'Jun', 'Sep', 'Nov'])) {
+                    $end_day = min($end_day, 30);
+                }
+
+                // Start from beginning of day these must be reported in GMT as logs are processed as such
+                $start_string = $_POST['start_month'] . ' ' . $start_day . ' ' . $_POST['start_year'] . ' 00:00:00';
```
Comment (levimeahan) [inline pages/modPanel.php:725]: Assuming someone used SQL injection on this field so the value of `$_POST['start_month']` is like `'; DELETE FROM users;--"`, I wouldn't expect DateTimeImmutable to pass through the text, but I don't know the internals, so let's sanitize to be safe: 
```suggestion
                $start_string = (int)$_POST['start_month'] . ' ' . $start_day . ' ' . (int)$_POST['start_year'] . ' 00:00:00';
```

File: pages/modPanel.php [hunk 4]
```diff
@@ -671,6 +677,162 @@ function modPanel() {
...
+                $start_time = new DateTimeImmutable($start_string, new DateTimeZone('Europe/London'));
+                // End just prior to midnight these must be reported in GMT as logs are processed as such
+                $end_string = $_POST['end_month'] . ' ' . $end_day . ' ' . $_POST['end_year'] . ' 23:59:59';
```
Comment (levimeahan) [inline pages/modPanel.php:728]:
```suggestion
                $end_string = (int)$_POST['end_month'] . ' ' . $end_day . ' ' . (int)$_POST['end_year'] . ' 23:59:59';
```

File: templates/staff/mod/ban_confirm.php [hunk 5]
```diff
@@ -13,13 +14,17 @@
         <td style="text-align: center;">
             <?php if($ban_length == StaffManager::PERM_BAN_VALUE): ?>
                 Issue permanent <?=ucwords($ban_type)?> Ban?
+            <?php elseif($ban_length >= StaffManager::MINUTES_PER_DAY * 30): ?>
+				<?=ucwords($ban_type)?> Ban <?=$user_data['user_name']?> for <?=($ban_length / (30 * StaffManager::MINUTES_PER_DAY))?> month(s)?<br />
```
Comment (levimeahan) [inline templates/staff/mod/ban_confirm.php:18]: Consider adding another constant for this just to make life a little easier
```suggestion
				<?=ucwords($ban_type)?> Ban <?=$user_data['user_name']?> for <?=($ban_length / StaffManager::MINUTES_PER_MONTH)?> month(s)?<br />
```

**Review by levimeahan — APPROVED** [review]

File: classes/StaffManager.php (hunk 2 above)
Comment (HitoriWoB) [reply classes/StaffManager.php:520]: I was thinking about doing this with the post-character staff panel upgrades that are needed. GMTA

File: pages/modPanel.php (hunk 3 above)
Comment (HitoriWoB) [reply pages/modPanel.php:725]: In my testing, malformed date-time strings failed to initialize.
I think we could move that down a bit further where the time stamps are pulled?

```php
                // Set stamps for queries
                $start_time_stamp = (int) $start_time->getTimestamp();
                $end_time_stamp = (int) $end_time->getTimestamp();
```

**Pull request merged**